test/params_api_test.c

```diff
@@ -0,0 +1,362 @@
+    params[n++] = OSSL_PARAM_construct_double("double", &d, &sz);
+    params[n++] = OSSL_PARAM_construct_BN("bignum", ubuf, sizeof(ubuf), &sz);
+    params[n++] = OSSL_PARAM_construct_utf8_string("utf8str", buf, sizeof(buf),
+                                                   &sz);
+    params[n++] = OSSL_PARAM_construct_octet_string("octstr", buf, sizeof(buf),
+                                                    &sz);
+    params[n++] = OSSL_PARAM_construct_utf8_ptr("utf8ptr", &bufp, &sz);
+    params[n++] = OSSL_PARAM_construct_octet_ptr("octptr", &vp, &sz);
+    memcpy(params + n, &pend, sizeof(params[n]));
+
+    /* Search failure */
+    if (!TEST_ptr_null(OSSL_PARAM_locate(params, "fnord")))
+        goto err;
+
+    /* All integral types */
+    for (j = 0; j < OSSL_NELEM(number_names); j++) {
+        if (!TEST_ptr(p = OSSL_PARAM_locate(params, number_names[j]))
+            || !TEST_true(OSSL_PARAM_set_uint32(p, (uint32_t)(3 + j)))
+            || !TEST_true(OSSL_PARAM_get_uint64(p, &u64))
+            || !TEST_size_t_eq(p->buffer_size, sz)
+            || !TEST_size_t_eq((size_t)u64, 3 + j)
+            || !TEST_true(OSSL_PARAM_get_double(p, &d2))
+            || !TEST_double_eq(d2, (double)u64)) {
+            TEST_note("iteration %zu var %s", j + 1, number_names[j]);
+            goto err;
+        }
+    }
+    /* Real */
+    if (!TEST_ptr(p = OSSL_PARAM_locate(params, "double"))
+        || !TEST_true(OSSL_PARAM_set_double(p, 3.14))
+        || !TEST_true(OSSL_PARAM_get_double(p, &d2))
+        || !TEST_size_t_eq(sz, sizeof(double))
+        || !TEST_double_eq(d, d2))
+        goto err;
+    /* UTF8 string */
+    bufp = NULL;
+    if (!TEST_ptr(p = OSSL_PARAM_locate(params, "utf8str"))
+        || !TEST_true(OSSL_PARAM_set_utf8_string(p, "abcdef"))
+        || !TEST_size_t_eq(sz, sizeof("abcdef"))
+        || !TEST_true(OSSL_PARAM_get_utf8_string(p, &bufp, 0))
+        || !TEST_str_eq(bufp, "abcdef"))
+        goto err;
+    OPENSSL_free(bufp);
+    bufp = buf2;
+    if (!TEST_true(OSSL_PARAM_get_utf8_string(p, &bufp, sizeof(buf2)))
+        || !TEST_str_eq(buf2, "abcdef"))
+        goto err;
+    /* UTF8 pointer */
+    bufp = buf;
+    sz = 0;
+    if (!TEST_ptr(p = OSSL_PARAM_locate(params, "utf8ptr"))
+        || !TEST_true(OSSL_PARAM_set_utf8_ptr(p, "tuvwxyz"))
+        || !TEST_size_t_eq(sz, sizeof("tuvwxyz"))
+        || !TEST_str_eq(bufp, "tuvwxyz")
+        || !TEST_true(OSSL_PARAM_get_utf8_ptr(p, &bufp2))
+        || !TEST_ptr_eq(bufp2, bufp))
+        goto err;
+    /* OCTET string */
+    vp = NULL;
+    if (!TEST_ptr(p = OSSL_PARAM_locate(params, "octstr"))
+        || !TEST_true(OSSL_PARAM_set_octet_string(p, "abcdefghi",
+                                                  sizeof("abcdefghi")))
+        || !TEST_size_t_eq(sz, sizeof("abcdefghi"))
+        || !TEST_true(OSSL_PARAM_get_octet_string(p, &vp, 0, &s))
+        || !TEST_size_t_eq(s, p->buffer_size)
```

Comment:
This can't be right.  The buffer is 100 chars, which is larger than the `sizeof("abcdefghi")`.  I think that you need to do the same kind of "cheating" as in the BN case above